pubsub/benchmarks_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pubsub_test
+
+import (
+	"context"
+	"flag"
+	"testing"
+
+	"gocloud.dev/gcp"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/gcppubsub"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+)
+
+var (
+	projectID = flag.String("benchmark-project", "", "project ID")
+	topicName = flag.String("benchmark-topic", "", "topic name")
+)
+
+func BenchmarkReceive(b *testing.B) {
+	if *projectID == "" || *topicName == "" {
+		b.Skip("missing -benchmark-project or -benchmark-topic")
+	}
+	attrs := map[string]string{"label": "value"}
+	body := []byte("hello, world")
+	const nMessages = 10000
+	topic, sub, cleanup, err := openGCPTopicAndSub()
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer cleanup()
+	b.Run("gcp", func(b *testing.B) {
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			b.StopTimer()
+			if err := publishN(topic, nMessages, 100, attrs, body); err != nil {
```

Comment:
Instead of 100, use `RunParallel`?
https://golang.org/pkg/testing/#B.RunParallel

Reply:
I don't quite see how I'd use it here. This is setup code; I want to publish a lot of messages so I can receive them in the timed part of the benchmark, and I want to do it quickly. I don't want to run the timed part of the benchmark in parallel, at least not here.

Reply:
Sorry, I meant more for the receiving part. Something like this:

```
    // publish b.N * nMessages messages
    b.ResetTimer
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			// receive nMessages messages
        }
    }
```

Maybe this doesn't work because it requires all `b.N * nMessages` messages to be published up front. But it lets the degree of concurrency/parallelism be determined by standard `go benchmark` flags, which is nice.

Reply:
Still doesn't make sense. `RunParallel` "creates multiple goroutines and distributes b.N iterations among them. " But `b.N` will aways be 1, because it takes long enough to run one iteration.